hybris: media: Make media_buffer_destroy delete the buffer

fredldotme opened the pull request:
Instead of lowering the reference count on the buffer (implicitly,
due to the Wrapper's dtor release() call) in the hopes of it getting
deleted, make sure that an explicit destroy is an explicit destroy.

Newer versions of Android apparently have refcounting bugs,
as noticed by LineageOS:
https://review.lineageos.org/c/LineageOS/android_frameworks_av/+/258972/2

Because of this the implicit release will not release the buffer at all.
Proposed solution: Have consumers of this API rely on checking the
refcount themselves and release/destroy accordingly.

Change-Id: I90afe51361c8453b4f7b63bfb02d327ee50a58ec

peat-psuwit commented:
I don't quite follow it.

See, the reason `mbuf->release()` was added in [1] is because `MediaCodecBuffer::setMediaBufferBase()` "stole" the reference, while `MediaBufferHolder::MediaBufferHolder()` doesn't. (And note that `MediaBufferBase` doesn't inherit from `RefBase` - but `MediaBufferHolder` does)

```c++
void MediaCodecBuffer::setMediaBufferBase(MediaBufferBase *mediaBuffer) {
      if (mMediaBufferBase != NULL) {
          mMediaBufferBase->release();
      }
      mMediaBufferBase = mediaBuffer;
  }
```

```c++
    MediaBufferHolder(MediaBufferBase* buffer)
        : mMediaBuffer(buffer) {
        if (mMediaBuffer != nullptr) {
            mMediaBuffer->add_ref();
        }
    }
```

And thus that's what LineageOS's changeset [2] is doing: call `release()` again after `MediaBufferHolder` has its.

What we're doing here is different from what [2] does. It sidesteps the ref-counting completely and do the explicit delete, which is IMO dangerous.

Opinion: `libmedia_compat_layer` shouldn't expose `MediaBufferBase`'s refcount as its own, since you can't distinguish between wrapper's references and references from somewhere else.

Maybe what's leaking is the wrapper and not the `MediaBufferBase` itself?

[1] https://android.googlesource.com/platform/frameworks/av/+/f03606d9034730bea1a394e6803f9ebc36f3d2eb%5E%21/#F13
[2] https://review.lineageos.org/c/LineageOS/android_frameworks_av/+/258972/2

peat-psuwit commented:
~~I think the problem lies here:~~

~~https://github.com/Halium/libhybris/blob/d5807df4945fb45831028c10f7946275c2d89c21/compat/media/media_buffer_layer.cpp#L288-L288~~

Aethercast doesn't even call that.

fredldotme commented:
If your proposed solution involves having to carry patches for every Halium version from 9 to 11 and beyond, diverging in the expected behavior from the rest of the system, then I'm not interested. Sorry.

Diverging as little as possible from Android would be a preferred goal in any case, since I don't want us to open a can of worms.

peat-psuwit commented:
I'm still trying to understand the problem. Do you know what creates/causes the "last reference"?

fredldotme commented:
> I'm still trying to understand the problem. Do you know what creates/causes the "last reference"?

The last reference is the observer that is assigned to the buffer. Apparently that one is not getting its onBufferReturned method called, even if the MediaBuffer is unref'ed yet another time.

fredldotme commented:
For the record, this is for working around an issue inside MediaCodecSource, which Aethercast is the only consumer of AFAICT.

The way I understand the situation wrt buffer handling:

- Newer Android introduced the observer concept to MediaBuffer (which MediaBufferHolder is one example of)
- The observer's reason for existance is to be notified when the MediaBuffer has its refcount down to 0 again and delete accordingly (in `onSignalBufferReturned`)
- Note this line here: https://android.googlesource.com/platform/frameworks/av/+/refs/heads/pie-gsi/media/libmediaextractor/MediaBuffer.cpp#99

My observation with an unpatched libmedia_compat_layer was that endlessly calling `release()` in a loop from the outside does not delete the MediaBuffer in question. And looking at `release()` it might be because the observer is still set, but the observer *also* doesn't get its buffer return handler called either, keeping the object around in memory.
The release is supposed to happen here: https://android.googlesource.com/platform/frameworks/av/+/refs/heads/pie-gsi/media/libstagefright/MediaCodecSource.cpp#431

This is what I'm a little confused about, since at least the observer would be in power to delete the object but doesn't do so.
So what this change proposes is overriding the observer after *we* (or rather Aethercast) are sure the buffer in question has its lifecycle completed, and destroy it (so basically what the buffer return handler is supposed to do) from the outside.
This also adds proper nesting of the object's lifecycle since `delete this` in MediaBuffer breaks the assumption that the outside has any control over it's child object's data.

peat-psuwit commented:
> My observation with an unpatched libmedia_compat_layer was that endlessly calling release() in a loop from the outside does not delete the MediaBuffer in question.

If we are the last reference holder of a `MediaBuffer`, endlessly calling `release()` in a loop should not be possible, because of a `CHECK()` condition [here](https://android.googlesource.com/platform/frameworks/av/+/refs/heads/pie-gsi/media/libmediaextractor/MediaBuffer.cpp#108). Ok, I might mis-remember (or [Chromium differs from Android](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#check_dcheck_and-notreached)) and `CHECK()` failure doesn't lead to a crash. If so, then it's possible that the the observer was never called because the `mRefCount` went below 0 (without calling `release()` in a loop of course) and the condition was strictly `(prevCount == 1)`.

So I reckon that there must be something wrong with reference count. I would say, having a breakpoint in a reference counting code (`release()` and `add_ref()`) could tell what happens and what's the value of the `mRefCount`.

fredldotme commented:
> > My observation with an unpatched libmedia_compat_layer was that endlessly calling release() in a loop from the outside does not delete the MediaBuffer in question.
> 
> If we are the last reference holder of a `MediaBuffer`, endlessly calling `release()` in a loop should not be possible, because of a `CHECK()` condition [here](https://android.googlesource.com/platform/frameworks/av/+/refs/heads/pie-gsi/media/libmediaextractor/MediaBuffer.cpp#108). Ok, I might mis-remember (or [Chromium differs from Android](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#check_dcheck_and-notreached)) and `CHECK()` failure doesn't lead to a crash. If so, then it's possible that the the observer was never called because the `mRefCount` went below 0 (without calling `release()` in a loop of course) and the condition was strictly `(prevCount == 1)`.
> 
> So I reckon that there must be something wrong with reference count. I would say, having a breakpoint in a reference counting code (`release()` and `add_ref()`) could tell what happens and what's the value of the `mRefCount`.

The observer was the one observing the MediaBuffer. "Reference counting" and "the observer" are two different things here. The observer might have a grab on the object, but it does not have an effect on counting the references itself. In that sense, he who owns the object only gets to clean up.

Gotta do the breakdance then, but I still think my way of accessing and deleting the MediaBuffer is a valid one, especially since the rest of StageFright assumes an observer to be optional. Things might change with future Android versions, but I would account for that in the consumer (Aethercast) or the compat library, not deep down in Android.

mariogrip commented:
I agree with @fredldotme that we should not change the android side for android bugs, its better to workaround this issue on our side of the table. But i would add `if observer is nullptr then only media_buffer_destroy delete d->buffer` that way we dont break anything else and it would be safe to do this. also then we specifically need to define that we allow to delete the buffers (without a api/abi change)

Also its better to have a fix that works, then no fix at all. and we don't have the time to *do everything* by the book, sure it might be *better* to fix every android version, but we also don't have the manpower to dig into these issue on every device as aeathercast is the only user right now.

peat-psuwit commented:
To be clear, when I said "set a breakpoint in reference counting code", I'm not saying that it's Android's fault. I'm still trying to understand what's actually going on. More likely than not, the problem lies in our libhybris. I don't believe that Android actually have a reference counting bug, only that our way of calling it is wrong.

Yes, a non-perfect solution is better than nothing. However, this PR feels like a way to shoot ourselves in the foot. If we really have to point our gun there, at least we should be sure we really need this. And at this point, I'm not convinced.

fredldotme commented:
... But *I am* convinced as it reproducibly fixes the issue at hand.

AFAICT the problem might be related to both libmedia_compat_layer & MediaCodecSource both trying to set an observer on the buffer at hand.
Sooo... in that sense, we've already shot ourselves in the foot, big time. Apparently we keep shooting ourselves in the foot from a teamwork perspective because we cannot move forward now.

"What if"s are a pain. Yes I see the concerns, of course it would be nice to have a future-proof solution all the time but the fact of the matter is:
Due to Android being a moving target we cannot guarantee it. And if you cannot guarantee it, brace yourself for having to make compromises. Trust me, been there, done that. And after all: libhybris compats are a mess of `ifdef`s anyway, we can bring order to chaos there should there ever be a need.

With that being said: I'm the only developer right now interfacing with this API, pretty much the primary maintainer of Aethercast and as I said: Nowhere in the Android tree does it say "observers are mandatory". Else they would be part of the respective ctors.
Hence I see it more than appropriate to allow memory management from the parents perspective.

fredldotme committed:
compat: media: Distinguish between managed & unmanaged MediaBuffers

The MediaCodecSource layer instantiate a MediaBufferPrivate directly with
a MediaBuffer provided, which most certainly comes from outside (like the
MediaCodecSource) and hence is supposed to be managed by that one using
another observer next to our own.

Change-Id: Icd343f3b38dc208c499cf71b0cb5e8e4a0e36861

fredldotme commented:
Alright, I found a different way, potentially worse, potentially better.

The solution emphasizes the problem: We're stuffing a MediaBuffer(Base) into a MediaBufferPrivate wrapper that
is managed by an outside entity, namely the MediaCodecSource inside libstagefright.

This makes the MediaBuffer(Base) inside the wrapper detached from destruction of the wrapper itself.

mariogrip commented:
> Alright, I found a different way, potentially worse, potentially better.
> 
> The solution emphasizes the problem: We're stuffing a MediaBuffer(Base) into a MediaBufferPrivate wrapper that is managed by an outside entity, namely the MediaCodecSource inside libstagefright.
> 
> This makes the MediaBuffer(Base) inside the wrapper detached from destruction of the wrapper itself.

If that works, it seems like a good way to do it. I would be happy to merge this then

fredldotme commented:
> > Alright, I found a different way, potentially worse, potentially better.
> > The solution emphasizes the problem: We're stuffing a MediaBuffer(Base) into a MediaBufferPrivate wrapper that is managed by an outside entity, namely the MediaCodecSource inside libstagefright.
> > This makes the MediaBuffer(Base) inside the wrapper detached from destruction of the wrapper itself.
> 
> If that works, it seems like a good way to do it. I would be happy to merge this then

Yup, works in the sense that memleaks are gone.

Flohack74 commented:
Please no merges without tracking it properly for next OTA! @mariogrip

JamiKettunen commented:
@Flohack74 Don't we need downstream package update anyway on https://gitlab.com/ubports/development/core/packaging/libhybris? I don't think we need to worry about UT things here :)

Actually I don't think I have the full picture, this probably integrates (only or also?) into the build Halium system images I'm guessing?

Flohack74 commented:
@JamiKettunen yes thats exactly what would happen, libhybris has a UT part and an Android part which is added to the manifests I believe.
